### PR TITLE
Proved PreferredRoundRule on the concrete side

### PR DESCRIPTION
**Pull request by lisandrasilva**
Proved `PreferredRoundRule` on the concrete side adding 2 new implementation obligations for this proof: `PR-ImplObligation₁` and `PR-ImplObligation₂`.  Besides that, we also modified the first implementation obligation of the `VotesOnceRule` to the new one `IncreasingRoundObligation`. We fixed the proof of `VotesOnce` using this new implementation obligation and also used it to prove the `PreferredRoundRule`.

### CODE REVIEW

File: LibraBFT/Concrete/Properties/PreferredRound.agda
Comment (mark-moir) [inline LibraBFT/Concrete/Properties/PreferredRound.agda:120]: You could simplify this by using `sameSig⇒sameVoteDataNoCol`, but it's no big deal because this proof is simple anyway, and hopefully one day we'll eliminate `meta-sha256-cr`, which will mean revisiting these kinds of proofs.

File: LibraBFT/Concrete/Properties/PreferredRound.agda
Comment (mark-moir) [inline LibraBFT/Concrete/Properties/PreferredRound.agda:115]: This and the next property are duplicated from `VotesOnce`.  Maybe factor them out into (say) `LibraBFT.Concrete.Properties.Common`?  Or at least add a `TODO-2` comment?

File: LibraBFT/Concrete/Properties/PreferredRound.agda
Comment (mark-moir) [inline LibraBFT/Concrete/Properties/PreferredRound.agda:178]: These four lines are repeated verbatim twice below, so it would be good to do some refactoring to reduce the redundancy.  One idea is to make a version of `α-ValidVote-trans` that takes a `VoteForRound∈` whose `msgVote` is the second `Vote`, rather than passing all the equalities, which are taken from such a `VoteForRound∈` anyway.  That makes the redundancy easier on the eye, but doesn't eliminate it.  Another possibility is to separate the two `with` abstractions at lines 184-187, so we first make the `v₁abs` in the `inj₂ v₁sb4` so we don't repeat it for both branches of the other one.

File: LibraBFT/Concrete/Properties/VotesOnce.agda
Comment (mark-moir) [inline LibraBFT/Concrete/Properties/VotesOnce.agda:61]: This is not used.  Instead we have an equivalent `postulate` called `¬genVotesRound≢0` in `LibraBFT.Impl.Properties.Handle`, which /is/ used in the concrete proofs (I think this is something I did at some point, so I take responsibility :smile:.)  The concrete proofs should not be reaching into a particular implementation for properties.  So I suggest removing the postulate mentioned above, and promoting `NewVoteSignedAndRound>0` to an implementation obligation.

File: LibraBFT/Impl/Properties/VotesOnce.agda
Comment (mark-moir) [inline LibraBFT/Impl/Properties/VotesOnce.agda:191]: I have completed this proof in another branch, so I will add it after you merge.

**Review by mark-moir — CHANGES_REQUESTED** [review]
Thanks @lisandrasilva, nice work!  I have a few comments that you could address in this branch.  However, there are already some conflicts and more are coming (see [this comment](https://github.com/oracle/bft-consensus-agda/pull/42#pullrequestreview-684535629)), so it may make sense to hold off and do a fresh pull request when the dust settles.

File: LibraBFT/Concrete/Properties/PreferredRound.agda
Comment (lisandrasilva) [reply LibraBFT/Concrete/Properties/PreferredRound.agda:120]: You are right, I just did it!

File: LibraBFT/Concrete/Properties/PreferredRound.agda
Comment (lisandrasilva) [reply LibraBFT/Concrete/Properties/PreferredRound.agda:115]: Yes, I was waiting for your opinion of what would be the most suitable place to put them. Thanks!

File: LibraBFT/Concrete/Properties/PreferredRound.agda
Comment (lisandrasilva) [reply LibraBFT/Concrete/Properties/PreferredRound.agda:178]: You are right! Fixed!

File: LibraBFT/Concrete/Properties/Common.agda
Comment (mark-moir) [inline LibraBFT/Concrete/Properties/Common.agda:1]: Please add a copyright notice.

File: LibraBFT/Concrete/Properties/VotesOnce.agda
Comment (mark-moir) [reply LibraBFT/Concrete/Properties/VotesOnce.agda]: @lisandrasilva are you going to address this one?

Comment (mark-moir) [reply LibraBFT/Concrete/Properties/VotesOnce.agda]: @lisandrasilva are you going to address this one?

**Review by mark-moir — CHANGES_REQUESTED** [review]
Looks good.  There is one change you didn't address, which I commented on, and one copyright notice to add.  After that, we will be merging other things that will cause even more conflicts (nothing difficult, I don't think), so hopefully you can do a fresh PR to supersede this one thereafter.  Thanks!

**Comment by mark-moir** [comment]
Great, thanks @lisandrasilva.  I have a few very minor comments, but I'm going to merge this and then address them myself in another PR later today, which will also include reinstating and fixing the commented-out proof in `LibraBFT.ImplFake.Properties.VotesOnce`.

**Pull request merged**
